File: ts-client/src/query.ts
import type {
  AggregateGroup,
  AggregateOp,
  AggregateSpec,
  FilterExpr,
  HybridSearchQuery,
  Order,
  PaginatedResultJson,
  QueryJson,
  SearchMode,
  SearchQuery,
  VectorQuery,
} from "./protocol.js";
import type { Doc, Id, IndexNamesOf, SchemaDefinition, TableNames } from "./schema.js";

/** A finished query carrying a phantom `Result` type used by the client/hooks. */
export interface RtQuery<Result> {
  readonly json: QueryJson;
  readonly __result?: Result;
}

/** Chainable builder for one table. `DocT` is the read-doc type; `Indexes` the index-name union. */
export class TableQuery<DocT, Indexes extends string> {
  constructor(private readonly json: QueryJson) {}

  /** Selects a declared index and an optional equality prefix (`eq`) over
   * its leading fields. Range bounds (`gt`/`gte`/`lt`/`lte`) apply to the
   * index field immediately after the `eq` prefix. */
  withIndex(index: Indexes, eq: unknown[] = []): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, index, eq });
  }

  /** Restricts the index field after the `eq` prefix to values strictly
   * greater than `value`. Requires `withIndex`. */
  gt(value: unknown): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, gt: value });
  }

  /** Restricts the index field after the `eq` prefix to values greater
   * than or equal to `value`. Requires `withIndex`. */
  gte(value: unknown): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, gte: value });
  }

  /** Restricts the index field after the `eq` prefix to values strictly
   * less than `value`. Requires `withIndex`. */
  lt(value: unknown): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, lt: value });
  }

  /** Restricts the index field after the `eq` prefix to values less than
   * or equal to `value`. Requires `withIndex`. */
  lte(value: unknown): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, lte: value });
  }

  /** Sets result order (`"asc"` or `"desc"`) by the selected index. */
  order(order: Order): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, order });
  }

  /** Append a db-side `filter` predicate. Composes with index/range/order/take;
   * the server validates terminal combinations. Not a terminal. */
  filter(expr: FilterExpr): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, filter: expr });
  }

  /** Field projection: each result doc keeps only the listed user fields plus
   * every `_`-prefixed key (the system fields `_id`/`_creationTime`/`_version`
   * and synthetics like `_searchSnippet` — always present; listing one is an
   * accepted no-op). Calling with no arguments (`fields()`) is the meaningful
   * system-fields-only (ids-only) view. Composes with every doc-bearing
   * terminal (get/collect/first/unique/paginate and the search family);
   * doc-less terminals (count/distinct/aggregate) are unaffected. Not a
   * terminal. Unknown names are rejected `BAD_REQUEST` by the server (and the
   * in-memory harness) at compile time. */
  fields(...names: string[]): TableQuery<DocT, Indexes> {
    return new TableQuery({ ...this.json, fields: names });
  }

  /** Full-text `search` over a declared search index. Composes only with `take`
   * (e.g. `.search("idx", "text").take(10)`); the server rejects every other
   * terminal alongside it. The optional `filter` narrows results server-side via
   * the full `FilterExpr` DSL (not to be confused with the query-level
   * `.filter()` builder, which is mutually exclusive with `search`); the
   * optional `mode` selects the match strategy — `tsquery` (default, word/stem
   * full-text) or `trgm` (case-insensitive substring, ranked by similarity);
   * the optional `snippet` (FM-31) opts each hit into a `_searchSnippet`
   * fragment with matched terms wrapped in `<mark>` (tsquery mode only) — all
   * omitted on the wire when absent so existing requests stay byte-identical. */
  search(
    index: string,
    query: string,
    opts?: { filter?: FilterExpr; mode?: SearchMode; snippet?: boolean },
  ): TableQuery<DocT, Indexes> {
    const search: SearchQuery = {
      index,
      query,
      ...(opts?.filter ? { filter: opts.filter } : {}),
      ...(opts?.mode ? { mode: opts.mode } : {}),
      ...(opts?.snippet !== undefined ? { snippet: opts.snippet } : {}),
    };
    return new TableQuery({ ...this.json, search });
  }

  /** Vector-similarity `vectorSearch` over a declared vector index. The server
   * ranks by the index's distance metric and applies `limit`; the optional
   * `filter` narrows results server-side via the full `FilterExpr` DSL (not to
   * be confused with the query-level `.filter()` builder); omitted on the wire
   * when absent so existing requests stay byte-identical. Terminal — the server
   * rejects other terminals alongside it. */
  vectorSearch(
    index: string,
    vector: number[],
    opts: { limit: number; filter?: FilterExpr },
  ): TableQuery<DocT, Indexes> {
    const vectorSearch: VectorQuery = {
      index,
      vector,
      limit: opts.limit,
      ...(opts.filter ? { filter: opts.filter } : {}),
    };
    return new TableQuery({ ...this.json, vectorSearch });
  }

  /** Hybrid `hybridSearch` terminal: fuses full-text and vector ranking over the
   * same table via Reciprocal Rank Fusion. The table must declare BOTH a search
   * index and a vector index. `opts.searchIndex`/`opts.vectorIndex` optionally
   * name the indexes (auto-selected when omitted); `opts.k` is the RRF constant
   * (default 60). Terminal — the server rejects other terminals alongside it. */
  hybridSearch(
    query: string,
    vector: number[],
    limit: number,
    opts?: { searchIndex?: string; vectorIndex?: string; k?: number },
  ): TableQuery<DocT, Indexes> {
    const hybridSearch: HybridSearchQuery = {
      query,
      vector,
      limit,
      ...(opts?.searchIndex ? { searchIndex: opts.searchIndex } : {}),
      ...(opts?.vectorIndex ? { vectorIndex: opts.vectorIndex } : {}),
      ...(opts?.k ? { k: opts.k } : {}),
    };
    return new TableQuery({ ...this.json, hybridSearch });
  }

  /** Terminal: returns up to `n` matching docs. */
  take(n: number): RtQuery<DocT[]> {
    return { json: { ...this.json, take: n } };
  }

  /** Terminal: returns the single matching doc, or `null` if none match.
   * The server rejects the query if more than one doc matches. */
  unique(): RtQuery<DocT | null> {
    return { json: { ...this.json, unique: true } };
  }

  /** Terminal: returns the first matching doc in the selected order, or
   * `null` if none match. */
  first(): RtQuery<DocT | null> {
    return { json: { ...this.json, first: true } };
  }

  /** Terminal: returns the count of matching docs. */
  count(): RtQuery<number> {
    return { json: { ...this.json, count: true } };
  }

  /** Distinct-values terminal: returns the unique values of the index field
   * immediately after the `eq` prefix over the matching set (an array of
   * scalar values, e.g. `["alice","bob"]`). Server rejects when no index is
   * set or the eq prefix consumes every index field; mutually exclusive with
   * every other terminal except `eq`/range bounds/`filter`. */
  distinct(): RtQuery<unknown[]> {
    return { json: { ...this.json, distinct: true } };
  }

  /** Aggregate terminal: runs `<op>` (SUM/AVG/MIN/MAX/COUNT) over the index
   * field immediately after the `eq` prefix. Without `groupBy`, returns one
   * scalar (`null` if no rows match; `0` for `count`). With `groupBy: true`,
   * groups by the index field after the eq prefix and aggregates the one after
   * that, returning `{key, value}[]` ordered by group key. `sum`/`avg` require
   * a numeric aggregate field; `count` aggregates rows and consumes no
   * aggregate field (a scalar `count` needs no index, a grouped `count` needs
   * one index field to group by). The server rejects non-numeric, no-index, or
   * no-field-beyond-prefix cases for the field-bearing ops. Mutually exclusive
   * with every other terminal except `eq`/range bounds/`filter` (which narrow
   * the matching set); `take` is also rejected — group count is capped
   * internally by MAX_TAKE. */
  aggregate(
    op: AggregateOp,
    groupBy: boolean = false,
  ): RtQuery<unknown> | RtQuery<AggregateGroup[]> {
    const spec: AggregateSpec = groupBy ? { op, groupBy: true } : { op };
    return { json: { ...this.json, aggregate: spec } };
  }

  /** Terminal: cursor-based pagination. `cursor` is the opaque cursor from
   * a prior page (`undefined`/empty starts from the beginning); `numItems`
   * caps the page size. Use `decodeCursor`/`encodeCursor` to inspect or
   * construct cursors. */
  paginate(cursor: string | undefined, numItems: number): RtQuery<PaginatedResultJson> {
    // ARC-133: Paginate.cursor is `?:`-optional, so include it only when set
    // (exactOptionalPropertyTypes forbids literal `undefined`). `cursor || ""`
    // collapses empty string to absence — same runtime as before.
    const c = cursor || "";
    return {
      json: {
        ...this.json,
        paginate: {
          ...(c === "" ? {} : { cursor: c }),
          numItems: numItems,
        },
      },
    };
  }

  /** Terminal: returns every matching doc (no limit). */
  collect(): RtQuery<DocT[]> {
    return { json: { ...this.json } };
  }
}

/** Per-table query surface returned by {@link createApi}. */
export interface TableApi<DocT, Indexes extends string> {
  /** Starts a new query builder for this table. */
  query(): TableQuery<DocT, Indexes>;
  /** Builds a query for the single doc with the given id, or `null` if
   * absent. */
  get(id: Id<string> | string): RtQuery<DocT | null>;
}

/**
 * Typed API client surface mapped from database schema definition S.
 */
export type ClientApi<S extends SchemaDefinition<any>> = {
  [T in TableNames<S>]: TableApi<Doc<S, T>, IndexNamesOf<S, T>>;
};

/** Builds a per-table typed query surface from a schema definition. */
export function createApi<S extends SchemaDefinition<any>>(schema: S): ClientApi<S> {
  const api: Record<string, TableApi<unknown, string>> = {};
  for (const table of Object.keys(schema.tables)) {
    api[table] = {
      query: () => new TableQuery<unknown, string>({ table }),
      get: (id: string) => ({ json: { table, get: id } }),
    };
  }
  return api as ClientApi<S>;
}

File: go-client/dsl/query.go
// go-client/dsl/query.go
package dsl

// Mirrors rust-client/src/query.rs::Query builder — a chainable value-type
// builder producing wire.Query.

import "github.com/paulrobello/par-rt-db/go-client/wire"

// TableQuery builds a wire.Query fluently.
type TableQuery struct {
	q wire.Query
}

// NewTableQuery starts a query on a table.
func NewTableQuery(table string) TableQuery {
	return TableQuery{q: wire.Query{Table: table}}
}

// Get targets a point read by id; excludes all range clauses.
func (q TableQuery) Get(id string) TableQuery {
	q.q.Get = &id
	return q
}

// WithIndex selects the index and binds the eq prefix values.
func (q TableQuery) WithIndex(name string, eq ...wire.JSONValue) TableQuery {
	q.q.Index = &name
	q.q.Eq = append(q.q.Eq, eq...)
	return q
}

// Eq appends one eq-prefix bind.
func (q TableQuery) Eq(v wire.JSONValue) TableQuery {
	q.q.Eq = append(q.q.Eq, v)
	return q
}

// Gt sets an exclusive lower bound.
func (q TableQuery) Gt(v wire.JSONValue) TableQuery {
	q.q.Gt = &v
	return q
}

// Gte sets an inclusive lower bound (mutually exclusive with Gt).
func (q TableQuery) Gte(v wire.JSONValue) TableQuery {
	q.q.Gte = &v
	return q
}

// Lt sets an exclusive upper bound.
func (q TableQuery) Lt(v wire.JSONValue) TableQuery {
	q.q.Lt = &v
	return q
}

// Lte sets an inclusive upper bound (mutually exclusive with Lt).
func (q TableQuery) Lte(v wire.JSONValue) TableQuery {
	q.q.Lte = &v
	return q
}

// OrderAsc orders ascending (the server default; explicit on the wire).
func (q TableQuery) OrderAsc() TableQuery {
	o := wire.OrderAsc
	q.q.Order = &o
	return q
}

// OrderDesc orders descending.
func (q TableQuery) OrderDesc() TableQuery {
	o := wire.OrderDesc
	q.q.Order = &o
	return q
}

// Take caps the result set.
func (q TableQuery) Take(n int) TableQuery {
	q.q.Take = &n
	return q
}

// Unique reads a single row via a unique index.
func (q TableQuery) Unique() TableQuery {
	q.q.Unique = true
	return q
}

// First returns Doc(Some)/Doc(None).
func (q TableQuery) First() TableQuery {
	q.q.First = true
	return q
}

// Count is the count terminal.
func (q TableQuery) Count() TableQuery {
	q.q.Count = true
	return q
}

// Distinct is the distinct terminal.
func (q TableQuery) Distinct() TableQuery {
	q.q.Distinct = true
	return q
}

// Aggregate is the aggregate terminal; groupBy shifts to grouped rows.
func (q TableQuery) Aggregate(op wire.AggregateOp, groupBy bool) TableQuery {
	q.q.Aggregate = &wire.AggregateSpec{Op: op, GroupBy: groupBy}
	return q
}

// PaginateCursor enables pagination with a prior cursor ("" = first page).
func (q TableQuery) PaginateCursor(cursor string, numItems int) TableQuery {
	p := wire.Paginate{NumItems: numItems}
	if cursor != "" {
		p.Cursor = &cursor
	}
	q.q.Paginate = &p
	return q
}

// Filter appends the db-side predicate.
func (q TableQuery) Filter(f wire.FilterExpr) TableQuery {
	q.q.Filter = f
	return q
}

// Search is the full-text search terminal.
func (q TableQuery) Search(s wire.SearchQuery) TableQuery {
	q.q.Search = &s
	return q
}

// Fields projects the named user fields.
func (q TableQuery) Fields(fields ...string) TableQuery {
	q.q.Fields = fields
	return q
}

// VectorSearch is the vector-similarity terminal.
func (q TableQuery) VectorSearch(v wire.VectorSearchQuery) TableQuery {
	q.q.VectorSearch = &v
	return q
}

// HybridSearch is the RRF-fused text+vector terminal.
func (q TableQuery) HybridSearch(h wire.HybridSearchQuery) TableQuery {
	q.q.HybridSearch = &h
	return q
}

// Build returns the wire shape (shallow copy).
func (q TableQuery) Build() wire.Query {
	return q.q
}
